### DISCUSSION TITLE
Testing router.query is undefined at first render

### DISCUSSION THREAD

**Discussion by Gw3i** [neuefische/questions#146] [Web] [answered]
Hello everyone!

I want to test my EditForm, but I'm getting this error: 
```shell
 Edit Form › should show error message when input is empty

    TypeError: Cannot read properties of null (reading 'query')

      17 | 	const router = useRouter();
      18 | 	const videos = useStore(state => state.videos);
    > 19 | 	const ref = router.query;
         | 	                   ^
```

I've goggled it and it may be a first rendering problem. So that in the first rendering the `query` is `undefined` and the second rendering is getting a value. Because the first rendering is returning `undefined` the test failing.

But I couldn't find a good solution how to fix my code that way the test passes.
I often saw this code as a solution but I'm not really sure how to implement it properly:
```javascript
const router = useRouter();
React.useEffect(() => {
  if (router.isReady) {
    // Code using query
    console.log(router.query);
   }
}, [router.isReady]);
````
Here is my PR: https://github.com/Gw3i/capstone-project/pull/45
And thats a code snippet with the problem:

```javascript
export default function EditForm() {
	const router = useRouter();
	const videos = useStore(state => state.videos);
	const ref = router.query;

	const setConfirmationMessage = useStore(state => state.setConfirmationMessage);
	const categories = useStore(state => state.categories);

	const videoQueryId = videos.filter(video => {
		return ref.id === video.id;
	});

	const onSubmit = data => {
		toBeEditedVideo.videoTitle = data.videoTitle;
		toBeEditedVideo.category = data.category;
		setConfirmationMessage('Great! The video was edited!');
		router.push('/');
	};

	const toBeEditedVideo = videoQueryId[0];
	const {
		register,
		handleSubmit,
		formState: { errors },
	} = useForm({ criteriaMode: 'all' });

```

Thank you for your time!

**Comment by pixelass** [comment, marked as answer]
`router.query` is not undefined but rather an empty object. `{}`.
The issue is, that you have to mock `useRouter` if you want to run it in tests, since jest does not run in next.js.

See: https://github.com/orgs/neuefische/discussions/110

Also your `useEffect` should look different.


```jsx
// Destructure the router object
const {query} = useRouter();
React.useEffect(() => {
    // Code using query
    console.log(query);
}, [query]);
```

**Reply by Gw3i** [reply]
Hey @pixelass 
thank you, that worked!
I don't even need the useEffect.
